### DISCUSSION TITLE
Texture replacements

### DISCUSSION THREAD

**Discussion by sultim-t** [sultim-t/xash-rt#124] [Announcements]
### Overview
![all](https://user-images.githubusercontent.com/31703959/222969613-d7568b8b-0ff0-4e54-96af-334cc8eeca51.png)

RayTracedGL1 accepts an unmodified texture from a game, and tries to replace it by searching for *a corresponding file*.

* If *not* devmode:
    * search for [KTX2](https://registry.khronos.org/KTX/specs/2.0/ktxspec.v2.html) file in the `rt/mat/` folder
    * if not found, use original data
* If devmode:
    * search for TGA / PNG / JPG file in the `rt/mat_dev/` folder
    * if not found, search for KTX2 file in the `rt/mat/` folder
    * if not found again, use original data
    
### About devmode
* In devmode, a developer menu is being run in a separate window

![image](https://user-images.githubusercontent.com/31703959/222971204-bfdaa3b8-fc58-4815-b273-6db7b2bb8bec.png)

* Files in the `rt/` folder are checked, and, if one was modified, RayTracedGL1 tries to hot-reload it
    * Note: but it doesn't hot-reload a texture, if its corresponding file didn't exist during its upload from a game :(
    
* To enable devmode:
   1. Create a text file named `RayTracedGL1.json` in the Half-Life folder
![image](https://user-images.githubusercontent.com/31703959/222972560-a696760a-5e0e-4b04-b0b0-cbb9b0bc7347.png)

   1. Put this text inside:
      ```
      {
          "version": 0,
          "developerMode": true,
          "vulkanValidation": false,
          "dlssValidation": false,
          "fpsMonitor": false
      }
      ```
    1. Launching `rt/xash3d.exe` should open an additional window named `RTGL1 Dev`
    
* To disable devmode:
    1. Either delete `RayTracedGL1.json`, or set `"developerMode": false`
    1. Launch a game

### Modifying textures in devmode
In Xash3D, there are multiple places in the code where a texture upload may happen. Usually, world textures have a name like: `#maps/sign50.mip`; animated models: `#models\barney\BX_Chrome1.mdl`; etc.

1. But it's easier just to export all textures via devmode: in `Textures` tab, click `Export original textures`

![image](https://user-images.githubusercontent.com/31703959/222970758-67b3090f-6f9b-4795-aa8f-5f44fdbb4526.png)

2. A folder `rt/mat_src` will be populated by all currently uploaded textures in TGA format. It may take some time

![image](https://user-images.githubusercontent.com/31703959/222971449-39c3b63c-c2c4-4754-b37e-9bf944642390.png)

3. For example, we would like to modify this sand texture:

![image](https://user-images.githubusercontent.com/31703959/222971672-789cbcfe-febe-4912-a34e-2ece1a547cb8.png)

4. Create `rt/mat_dev` folder and copy the file into it. The file must **preserve the file hierarchy** it previously had in `rt/mat_src`, e.g. the target file for *"rt/**mat_src**/#maps/out_snd2.mip.tga"* is *"rt/**mat_dev**/#maps/out_snd2.mip.tga"*

![image](https://user-images.githubusercontent.com/31703959/222971900-df7b86b0-1a0a-49d8-a29c-50e0f3886f88.png)

5. Modify the texture:

![image](https://user-images.githubusercontent.com/31703959/222972233-3cacf2d6-ce2c-4c75-9ee7-a100592fb032.png)

6. On map restart we should see this (given that devmode is enabled):

![image](https://user-images.githubusercontent.com/31703959/222972292-6aac5ca1-1701-4581-b16e-5dcc9c06172a.png)

### Modified textures in non-devmode

When devmode is disabled, only KTX2 textures can be used, as they are parsed in a fast manner and can be copied directly to a video memory wihout preprocessing. But this also means that textures need to be converted from TGA / PNG / etc to KTX2 before launching a game.

A small script `rt/CreateKTX2.py` converts TGA and PNG files from `rt/mat_dev/` folder to KTX2 files in `rt/mat/` folder. It also creates a `CreateKTX2Cache.txt` file to track what files have changed to prevent re-exporting the same data.

To use the script:
* install Python3 
* install [GPUOpen Compressonator](https://github.com/GPUOpen-Tools/compressonator/releases)
   * this is a tool that converts to KTX2
* a path to its `compressonatorcli.exe` must be added to PATH environment variable
    * so a `compressonatorcli.exe` can be found and called from any directory (as the script just calls that .exe)
    
After that, launching `CreateKTX2.py` should convert textures to KTX2, and we would be able to see the changed textures even with disabled devmode.

### PBR info

Textures can have additional images that define properties for physically-based rendering:
* metallic, roughness *(`_orm` suffix)*
* emission *(`_e` suffix)*
* normal map *(`_n` suffix)*

For example, we would like *"#models/v_crowbar/crowbar.mdl"* to be more metallic:

* enable devmode (just for faster iterations and hot-reloading)

* create a metallic-roughness texture
    * Red channel -- unused
    * Green channel -- roughness
    * Blue channel -- metallic
    * more info in the GLTF2 standard: [[1]](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#materials) [[2]](https://github.com/KhronosGroup/glTF-Tutorials/blob/master/gltfTutorial/gltfTutorial_010_Materials.md)

* save to a file called *"mat_dev/#models/v_crowbar/crowbar.mdl_orm.tga"*
    * notice the `_orm` suffix
![image](https://user-images.githubusercontent.com/31703959/222975882-8c989fd2-a655-4615-a9a7-4ed7ee3af72c.png)

* if a file **didn't exist** when original texture was uploaded, force this texture to be uploaded by a game, which happens for most textures on connecting to a map:
    * just **quick-save, type `disconnect` in the console and load the save**
    * this needs to be done as RayTracedGL1 doesn't check unknown files for hot-reload
    * this "re-uploading" technique should be done only **once** and only for **new files**, because after that RayTracedGL1 already knows about the files

* now PBR properties should be visible in the game

![image](https://user-images.githubusercontent.com/31703959/222977612-12ca2daa-8fc7-4c4a-8af0-6414b204d336.png)

Same process can be applied to:
* Emissive
    * with `_e` suffix
    * RGB represents emissive color [[reference]](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#additional-textures)
* Normal maps
    * with `_n` suffix
    * normal direction offset from a geometry normal [[reference]](https://registry.khronos.org/glTF/specs/2.0/glTF-2.0.html#additional-textures)
        * Red - offset along tangent
            * 0..255 remapped to -1..1
        * Green - offset along bitangent
            * 0..255 remapped to -1..1
        * Blue ignored

Note: same process for converting to KTX2 needs to be done

Note: to check which PBR textures are active, look `Textures` tab of the devmode window
    * A - albedo
    * P - metallic / roughness
    * N - normal map
    * E - emissive
    * each number is an internal texture ID
![image](https://user-images.githubusercontent.com/31703959/222977095-c7d67b9a-fd59-4a41-89f4-01413f3edfe7.png)

### Texture properties

There is some limited amount of texture properties that can be applied to a texture by its path.

For example: 
* we have found that *"mat_src/#models/glassgibs/chome_glass.mdl.tga"* should be glass
* so we need to set some properties to a texture with name *"#models/glassgibs/chome_glass.mdl"*
    * notice that `.tga` (and `mat/src/`) were removed, so we have the original texture name
        * be careful with extensions!
    * original texture names can be also found in `Primitives` / `Textures` tabs of devmode window (use middle mouse click to copy name, or right click -> Copy texture name)
        
![image](https://user-images.githubusercontent.com/31703959/222974607-bde0b5b9-345f-4065-8ce7-eb1c40c910ff.png)
        
* add an entry to `array` in `rt/data/textures.json`:
     * `{"textureName":"#models/glassgibs/chome_glass.mdl","isGlass":true}`
     
* save JSON
     * if in devmode, it should be hot-reloaded, and the result should be observed in the running game
     * otherwise, a map change or game restart is required

*Warning:* JSON parser has some bugs related to spaces inside arrays, specifically, there must not be any spaces around colons `:`
*Note:* Check the devmode's `Log` tab for parsing errors, after saving JSON files

A list of texture properties:
* textureName
* forceIgnore
* forceIgnoreIfRasterized
* forceAlphaTest
* forceTranslucent
* forceOpaque
* forceGenerateNormals
* forceExactNormals
* isMirror
* isWater
* isWaterIfTranslucent
* isGlass
* isGlassIfTranslucent
* isAcid
* isGlassIfSmooth
    * glass if roughness is small
* isMirrorIfSmooth
    * mirror if roughness is small
* isThinMedia
* metallicDefault -- default metallicity 
* roughnessDefault
* emissiveMult
* lightIntensity
    * a float that defines intensity of a light source that should be attached to a polygon with the given texture
* lightColor
    * attached light's color in 0..255, e.g. [255, 0, 0] defines a red color
* lightEvenOnDynamic
    * if false (default), lights will be attached to textured polygons *only* on geometry export
    * if true, lights will be attached to any geometry (both exportable and non-exportable); use with care: might increase amount of noise significantly

**Comment by magpie-o** [comment]
@sultim-t We will very appreciate if you will answer to many people on SYS_CRASH problem

**Reply by sultim-t** [reply]
Again, I'm out of ideas for that sys_crash. Someone who has the problem needs to debug it, as I am not sure if I can reproduce it on my machine :(

**Reply by magpie-o** [reply]
Understood. Thank you very much for your reply!

**Reply by Muxinio** [reply]
the error occurs if you have EVGA or MSI AFTERBOUNER open. I don't know if it could also be something else, I closed it and it didn't happen again.

**Comment by BryanAM7** [comment]
How can i adapt the og crowbar from 1998? i just copied the v_crowbar.mdl from  "\Half-Life\valve" to "Half-Life\rt\valve\models" and the crowbar looks fantastic but the reflexions update slow like at 1 fps or less so should i extract and modify the original textures or just add pbr propeties to them.

**Comment by RiceWitch** [comment]
how do you get devmode?!??!
